### PR TITLE
New feature: Direct LED Matrix

### PR DESCRIPTION
**Pull request by skullydazed**
## Description

This adds a new type of driver to led_matrix- LEDs directly connected to the MCU. It supports LEDs connected directly to pins and an LED matrix with rows and columns.

## Types of Changes

- [x] Core
- [x] New feature
- [x] Keyboard (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by skullydazed** [comment]
My latest commit does something new in QMK- it moves some tasks to a separate thread. Without this I get poor matrix scanning performance on stm32f303, such that I have to hold keys down at least 50ms. With this I can press keys as fast as possible.

**Comment by zvecr** [comment]
You could try what I have found quite accurate for ARM split serial, which i have been using a logic analyser to verify timings with.

```c
#define wait_us(x) chSysPolledDelayX(US2RTC(STM32_SYSCLK, x))
```

**Comment by drashna** [comment]
Been using this a bit on my Planck EZ, which uses the rgb matrix, and it works well!
`send_string` no longer halts the animations. 

And I don't see any other issues, though, this hasn't been an in depth test.

**Comment by skullydazed** [comment]
This PR is getting... large.

I ran into contention with the separate thread, so for now I'm not going with that approach. I may open that in a separate branch to continue experimenting with.

I'm going to try to split this PR into a few different PR's that are easier to manage.

**Comment by skullydazed** [comment]
Ok, after these PR's are merged I'll be able to simplify this PR a lot by rebasing:

* #7124 
* #7125 
* #7127

**Comment by skullydazed** [comment]
OK, this PR is ready for review. The two keyboards that use it have been merged so the only keyboard changes in this PR are related to enabling this new system.

File: quantum/stm32/proton_c.mk [hunk 1]
```diff
@@ -3,7 +3,10 @@
 # These are defaults based on what has been implemented for ARM boards
 AUDIO_ENABLE = yes
 RGBLIGHT_ENABLE = no
-BACKLIGHT_ENABLE = no
+
+ifneq ($(strip $(LED_MATRIX_ENABLE)), direct)
```
Comment (zvecr) [inline quantum/stm32/proton_c.mk]: what scenario does this cover? LED_MATRIX_ENABLE = direct doesnt seem supported then it turns backlight off?

Comment (skullydazed) [reply quantum/stm32/proton_c.mk]: Ah shoot, forgot to update that section when that changed. I may not need this now, I'll have to revisit this aspect.

**Comment by skullydazed** [comment]
The whitespace looks like an artifact of collapsing my commits down. I'll back those out.

**Comment by r2d2rogers** [comment]
I was able to use the direct pins feature on a bpiphany/sixshooter after advice from @zvecr to change the led_wait_us define:
https://github.com/r2d2rogers/qmk_firmware/blob/directLed/drivers/led/led_matrix_pins.c#L23-L24

Other changes for the sixshooter that were not in the feature documentation:
Add BACKLIGHT_LEVELS
Correct _user functions to _keymap functions for userspace

The sixshooter uses a Teensy 2.0

**Comment by smp4488** [comment]
I am very interested in this. I’m going to pull the changes into my branch tonight and see if it works.

**Comment by tzarc** [comment]
Removed "on hold" so stalebot can eventually close this.

**Pull request closed without merging**